keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,641 @@
+# KEP-83: Workload preemption
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Contraints/Caveats (Optional)](#notescontraintscaveats-optional)
+    - [How to configure waitBeforePreemptionSeconds](#how-to-configure-waitbeforepreemptionseconds)
+    - [Why no control to opt-out a ClusterQueue from preemption](#why-no-control-to-opt-out-a-clusterqueue-from-preemption)
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Workload preemption doesn't imply immediate Pod termination](#workload-preemption-doesnt-imply-immediate-pod-termination)
+    - [No mechanism to fail or terminate a Job](#no-mechanism-to-fail-or-terminate-a-job)
+  - [Increased admission latency](#increased-admission-latency)
+- [Design Details](#design-details)
+  - [API](#api)
+    - [ClusterQueue changes](#clusterqueue-changes)
+  - [Workload changes](#workload-changes)
+  - [Changes in scheduling algorithm](#changes-in-scheduling-algorithm)
+    - [Detecting Workloads that might benefit from preemption](#detecting-workloads-that-might-benefit-from-preemption)
+    - [Sorting among heads of ClusterQueues](#sorting-among-heads-of-clusterqueues)
+    - [Admission](#admission)
+    - [Preemption](#preemption)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Inform how costly is to interrupt a Workload](#inform-how-costly-is-to-interrupt-a-workload)
+  - [Extra knobs in ClusterQueue preemption policy](#extra-knobs-in-clusterqueue-preemption-policy)
+<!-- /toc -->
+
+## Summary
+
+This enhancement introduces workload preemption, a mechanism to suspend
+workloads when:
+- ClusterQueues under their minimum quota need the resources that are currently
+  lent to other ClusterQueues in the cohort.
+- Within a Clusterqueue, there are running Workloads with lower priority than
+  a pending Workload.
+
+API fields in the Workload and ClusterQueue definitions determine preemption policies.
+
+## Motivation
+
+When ClusterQueues under their minimum quota lend resources, they should
+be able to recover those resources fast, to be able to admit Workloads
+when there are sudden spikes. Similarly, the ClusterQueue should be able to
+recover quota from low priority workloads that are currently running.
+
+Currently, the only mechanism to recover those resources is to wait for
+Workloads to finish, which is generally unbounded.
+
+### Goals
+
+- Preempt Workloads from ClusterQueues borrowing resources when other
```

Comment:
Will we consider priority here?  E.g. 
1. We have two clusterQueues, cqA and cqB. They belong to a same cohort.
3. We create a workloadA in cqA, but cqA is insufficient of resources, so it will lend some resources from cqB
4. Then we create a workloadB in cqB, workloadB's priority is lower than workloadA, but cqB is also lack of resources, so at the moment:
    1）we'll perform preemption although workloadA's priority is higher. It shows that I can share some resources with neighbors, but when I need, I will reclaim.
    2)  we will not preempt workloadA for it's priority is higher, we respect priority whatever.

This is relevant to the idea I mentioned here https://github.com/kubernetes-sigs/kueue/pull/410#discussion_r997125757, if we support option1, this can help in same cases, like with hybrid orchestration workloads,  online ones are more sensitive, they should not be reclaimed.